### PR TITLE
GPU-soil-only-for-debug(ck)

### PR DESCRIPTION
**Pull request by LZUlis**
I am currently working on the development of soil mechanics for the CK version. At the moment, the code can only run correctly in debug mode. In ReleaseWithDefinfo mode, it can build successfully, but the runtime results are incorrect, with the time step becoming -NaN. In Release mode, it fails to build altogether.

I have tried various solutions found online, including adding "CMAKE_CXX_FLAGS": "-fsanitize=address", displaying the number of neighbors, and more, but none of them have resolved the issue. 

In the version demonstrated last Friday, there was an error in ReleaseWithDefinfo mode, which appeared to be caused by an out-of-bounds index. I later identified the root cause: the configuration (soil_block_update_complex_relation.exec()) was not initialized. However, even though the current version can run correctly in debug mode under ReleaseWithDefinfo, the time step still becomes -NaN, and I have made little progress in resolving this issue.

I am seeking assistance with this problem.

### CODE REVIEW

**Comment by Xiangyu-Hu** [comment]
@LZUlis 

My Comments are inline:

> I am currently working on the development of soil mechanics for the CK version. At the moment, the code can only run correctly in debug mode. In ReleaseWithDefinfo mode, it can build successfully, but the runtime results are incorrect, with the time step becoming -NaN. In Release mode, it fails to build altogether.

What is the error message when building in Release mode?
It may leads to the bug.

> 
> I have tried various solutions found online, including adding "CMAKE_CXX_FLAGS": "-fsanitize=address", displaying the number of neighbors, and more, but none of them have resolved the issue.
> 
> In the version demonstrated last Friday, there was an error in ReleaseWithDefinfo mode, which appeared to be caused by an out-of-bounds index. I later identified the root cause: the configuration (soil_block_update_complex_relation.exec()) was not initialized. 

What do you mean "not initialized"? not run?

However, even though the current version can run correctly in debug mode under ReleaseWithDefinfo, 

If it run correctly, how can the step become NaN?

the time step still becomes -NaN, and I have made little progress in resolving this issue.
> 
> I am seeking assistance with this problem.

**Comment by LZUlis** [comment]
Thanks very much for your review.

I have checked and there are the detailed description:

**1. In ck version.**
1.1 Debug mode
Bluid √， Launch √, Results √
1.2 Release/ReleasewithDefinfo mode
Build √， Launch √， Results ×
There are the results:
The acoustic_dt becomes -NAN:
image



**2. In sycl version**
2.1 Debug mode
Build ×
There are the build errors:
[build] ptxas /tmp/collumn_sycl-sm_50-8e8298-3016a9.s, line 149403; fatal   : Parsing error near '.': syntax error
[build] ptxas fatal   : Ptx assembly aborted due to errors
[build] llvm-foreach: 
[build] icpx: error: ptxas command failed with exit code 255 (use -v to see invocation)
[build] Intel(R) oneAPI DPC++/C++ Compiler 2024.0.0 (2024.0.0.20231017)
[build] Target: x86_64-unknown-linux-gnu
[build] Thread model: posix
[build] InstalledDir: /opt/intel/oneapi/compiler/2024.0/bin/compiler
[build] Configuration file: /opt/intel/oneapi/compiler/2024.0/bin/compiler/../icpx.cfg
[build] icpx: note: diagnostic msg: Error generating preprocessed source(s).
[build] gmake[3]: *** [tests/tests_sycl/2d_examples/test_2d_collumn_sycl/CMakeFiles/test_2d_collumn_sycl.dir/build.make:118：tests/tests_sycl/2d_examples/test_2d_collumn_sycl/bin/test_2d_collumn_sycl] 错误 1
[build] gmake[2]: *** [CMakeFiles/Makefile2:5604：tests/tests_sycl/2d_examples/test_2d_collumn_sycl/CMakeFiles/test_2d_collumn_sycl.dir/all] 错误 2
[build] gmake[1]: *** [CMakeFiles/Makefile2:5611：tests/tests_sycl/2d_examples/test_2d_collumn_sycl/CMakeFiles/test_2d_collumn_sycl.dir/rule] 错误 2
[build] gmake: *** [Makefile:1409：test_2d_collumn_sycl] 错误 2
[proc] The command: /usr/local/bin/cmake --build /home/li/Myfork/SPHinXsys-GPU_soil/build --parallel 130 --target test_2d_collumn_sycl -- exited with code: 2
_Note: there are smae erros in dambreack_sycl._

2.2 Release/ReleasewithDefinfo  mode

The acoustic_dt becomes too large:
image
_Note: the dambreak_sycl can run successfully and obtain correct results._

I'm really confused, and thanks for your review and help

**Comment by LZUlis** [comment]
> @LZUlis
> 
> My Comments are inline:
> 
> > I am currently working on the development of soil mechanics for the CK version. At the moment, the code can only run correctly in debug mode. In ReleaseWithDefinfo mode, it can build successfully, but the runtime results are incorrect, with the time step becoming -NaN. In Release mode, it fails to build altogether.
> 
> What is the error message when building in Release mode? It may leads to the bug.
> 
> > I have tried various solutions found online, including adding "CMAKE_CXX_FLAGS": "-fsanitize=address", displaying the number of neighbors, and more, but none of them have resolved the issue.
> > In the version demonstrated last Friday, there was an error in ReleaseWithDefinfo mode, which appeared to be caused by an out-of-bounds index. I later identified the root cause: the configuration (soil_block_update_complex_relation.exec()) was not initialized.
> 
> What do you mean "not initialized"? not run?
> 
> However, even though the current version can run correctly in debug mode under ReleaseWithDefinfo,
> 
> If it run correctly, how can the step become NaN?
> 
> the time step still becomes -NaN, and I have made little progress in resolving this issue.
> 
> > I am seeking assistance with this problem.
> 
> Please try to be as accurate and informative as possible when discuss technical issues.

Thanks for your reply and I have uploaded a new comment in this PR to describe my issues, Please review it if your are free.

**Comment by Xiangyu-Hu** [comment]
> Thanks very much for your review.
> 
> I have checked and there are the detailed description:
> 
> **1. In ck version.** 1.1 Debug mode Bluid √， Launch √, Results √ 1.2 Release/ReleasewithDefinfo mode Build √， Launch √， Results × There are the results: The acoustic_dt becomes -NAN: image
> 
> **2. In sycl version** 2.1 Debug mode Build × There are the build errors: [build] ptxas /tmp/collumn_sycl-sm_50-8e8298-3016a9.s, line 149403; fatal : Parsing error near '.': syntax error [build] ptxas fatal : Ptx assembly aborted due to errors [build] llvm-foreach: [build] icpx: error: ptxas command failed with exit code 255 (use -v to see invocation) [build] Intel(R) oneAPI DPC++/C++ Compiler 2024.0.0 (2024.0.0.20231017) [build] Target: x86_64-unknown-linux-gnu [build] Thread model: posix [build] InstalledDir: /opt/intel/oneapi/compiler/2024.0/bin/compiler [build] Configuration file: /opt/intel/oneapi/compiler/2024.0/bin/compiler/../icpx.cfg [build] icpx: note: diagnostic msg: Error generating preprocessed source(s). [build] gmake[3]: *** [tests/tests_sycl/2d_examples/test_2d_collumn_sycl/CMakeFiles/test_2d_collumn_sycl.dir/build.make:118：tests/tests_sycl/2d_examples/test_2d_collumn_sycl/bin/test_2d_collumn_sycl] 错误 1 [build] gmake[2]: *** [CMakeFiles/Makefile2:5604：tests/tests_sycl/2d_examples/test_2d_collumn_sycl/CMakeFiles/test_2d_collumn_sycl.dir/all] 错误 2 [build] gmake[1]: *** [CMakeFiles/Makefile2:5611：tests/tests_sycl/2d_examples/test_2d_collumn_sycl/CMakeFiles/test_2d_collumn_sycl.dir/rule] 错误 2 [build] gmake: *** [Makefile:1409：test_2d_collumn_sycl] 错误 2 [proc] The command: /usr/local/bin/cmake --build /home/li/Myfork/SPHinXsys-GPU_soil/build --parallel 130 --target test_2d_collumn_sycl -- exited with code: 2 _Note: there are smae erros in dambreack_sycl._
> 
> 2.2 Release/ReleasewithDefinfo mode
> 
> The acoustic_dt becomes too large: image _Note: the dambreak_sycl can run successfully and obtain correct results._
> 
> I'm really confused, and thanks for your review and help

We must make it works for CK mode before try sycl.

**Comment by Xiangyu-Hu** [comment]
> Thanks very much for your review.
> 
> I have checked and there are the detailed description:
> 
> **1. In ck version.** 1.1 Debug mode Bluid √， Launch √, Results √ 1.2 Release/ReleasewithDefinfo mode Build √， Launch √， Results × There are the results: The acoustic_dt becomes -NAN: image
> 
> **2. In sycl version** 2.1 Debug mode Build × There are the build errors: [build] ptxas /tmp/collumn_sycl-sm_50-8e8298-3016a9.s, line 149403; fatal : Parsing error near '.': syntax error [build] ptxas fatal : Ptx assembly aborted due to errors [build] llvm-foreach: [build] icpx: error: ptxas command failed with exit code 255 (use -v to see invocation) [build] Intel(R) oneAPI DPC++/C++ Compiler 2024.0.0 (2024.0.0.20231017) [build] Target: x86_64-unknown-linux-gnu [build] Thread model: posix [build] InstalledDir: /opt/intel/oneapi/compiler/2024.0/bin/compiler [build] Configuration file: /opt/intel/oneapi/compiler/2024.0/bin/compiler/../icpx.cfg [build] icpx: note: diagnostic msg: Error generating preprocessed source(s). [build] gmake[3]: *** [tests/tests_sycl/2d_examples/test_2d_collumn_sycl/CMakeFiles/test_2d_collumn_sycl.dir/build.make:118：tests/tests_sycl/2d_examples/test_2d_collumn_sycl/bin/test_2d_collumn_sycl] 错误 1 [build] gmake[2]: *** [CMakeFiles/Makefile2:5604：tests/tests_sycl/2d_examples/test_2d_collumn_sycl/CMakeFiles/test_2d_collumn_sycl.dir/all] 错误 2 [build] gmake[1]: *** [CMakeFiles/Makefile2:5611：tests/tests_sycl/2d_examples/test_2d_collumn_sycl/CMakeFiles/test_2d_collumn_sycl.dir/rule] 错误 2 [build] gmake: *** [Makefile:1409：test_2d_collumn_sycl] 错误 2 [proc] The command: /usr/local/bin/cmake --build /home/li/Myfork/SPHinXsys-GPU_soil/build --parallel 130 --target test_2d_collumn_sycl -- exited with code: 2 _Note: there are smae erros in dambreack_sycl._
> 
> 2.2 Release/ReleasewithDefinfo mode
> 
> The acoustic_dt becomes too large: image _Note: the dambreak_sycl can run successfully and obtain correct results._
> 
> I'm really confused, and thanks for your review and help

When the time step becomes NAN? do you have any intermediate results before it?

**Comment by Xiangyu-Hu** [comment]
@LZUlis we will not be able to use debug mode for sycl code. That is why we do the CK mode first.

File: tests/tests_sycl/2d_examples/test_2d_column_collapse_sycl/column_collapse_sycl.cpp [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+            Real relaxation_time = 0.0;
+            Real acoustic_dt = 0.0;
+            while (relaxation_time < Dt)
+            {
+                acoustic_dt = soil_acoustic_time_step.exec();
+                soil_acoustic_step_1st_half.exec(acoustic_dt);
+                soil_acoustic_step_2nd_half.exec(acoustic_dt);
+                relaxation_time += acoustic_dt;
+                integration_time += acoustic_dt;
+                sv_physical_time->incrementValue(acoustic_dt);
```
Comment (Xiangyu-Hu) [inline tests/tests_sycl/2d_examples/test_2d_column_collapse_sycl/column_collapse_sycl.cpp:193]: Could you show me the debug results for this steps? That is how the variables changes their values after each step?

Comment (Xiangyu-Hu) [reply tests/tests_sycl/2d_examples/test_2d_column_collapse_sycl/column_collapse_sycl.cpp:193]: he line acoustic_dt = soil_acoustic_time_step.exec(); is responsible for computing the time step acoustic_dt for the acoustic step. If acoustic_dt goes to NAN (Not a Number), it indicates that there is likely an issue in the calculation within the soil_acoustic_time_step.exec() function. Here are some possible reasons for this:

    Invalid Initial Conditions: The initial conditions or parameters might be set incorrectly, leading to invalid calculations.
    Division by Zero: There could be a division by zero occurring within the soil_acoustic_time_step.exec() function.
    Uninitialized Variables: Some variables used in the calculation might not be initialized properly.
    Numerical Instabilities: The numerical method used might be causing instabilities, especially if the time step is too large or if the material properties are not set appropriately.
    Incorrect Parameter Values: Parameters such as material properties (density, modulus, etc.) might be set to unrealistic values, leading to invalid results.

To debug this issue, you can:

    Check the initial values and parameters used in the simulation.
    Add debug statements to trace the values of variables within the soil_acoustic_time_step.exec() function.
    Ensure that all variables are properly initialized before they are used in calculations.
    Verify that the numerical methods and parameters are appropriate for the simulated scenario.

**Pull request closed without merging**